ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -717,6 +728,74 @@ impl fmt::Display for AddInstruction {
     }
 }
 
+/// The operand for a [Instruction::Icmp]
+///
+/// # Semantics
+///
+/// Compares two operands.
```

Comment:
Not sure actually. I think it could also be floating points: https://llvm.org/doxygen/classllvm_1_1CmpInst.html#a7bd850492d35a34f0fe419e5555997a9

Reply:
That'd be a [`fcmp`](https://llvm.org/docs/LangRef.html#fcmp-instruction) in LLVM.

[`icmp`](https://llvm.org/docs/LangRef.html#icmp-instruction) is for integers.

I assume we will copy this idiom?

Reply:
Ah, I see. `icmp` can still take a vec of ints though.

Reply:
Yeah, it's not uncommon for llvm instructions to have vectored modes.

Really we should be asserting that we don't see them when lowering our IR.

Reply:
Why would be assert this? Are we not expecting to seen them ever?

Reply:
There's no reason they couldn't appear in the AOT LLVM IR, right? I bet there are ways to get clang to generate that kind of code.

Reply:
Yes, so why would be assert this then? We should deal with those cases in the codegen? It should probably be a todo in the codegen, no?

Reply:
Because I doubt we lower them properly currently. Let's consider this a separate issue.

What I'm interested in here is, is it valid in our IR to compare an i32 with a u64, for example?

This is invalid in LLVM IR:

> They must also be identical types.

So I suppose it's easiest to inherit that for our IR too? In which case I think we should assert the operand types are the same.

Reply:
So I guess we would assert this in the codegen too?

Reply:
I think in this case the argument from the other comment applies here too: asserting this in the codegen means we cover both handwritten AOT IR and handwritten traces (while asserting this in the trace builder only covers AOT IR).

Reply:
ok, consider this resolved.